S4TF_Tutorial_3.ipynb

```diff
@@ -0,0 +1,1772 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "DiakU6qi_t57",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "/// Reads a file into an array of bytes.\n",
+        "func readFile(_ path: String) -> [UInt8] {\n",
+        "    let possibleFolders  = [\".\", \"MNIST\"]\n",
+        "    for folder in possibleFolders {\n",
+        "        let parent = URL(fileURLWithPath: folder)\n",
+        "        let filePath = parent.appendingPathComponent(path)\n",
+        "        guard FileManager.default.fileExists(atPath: filePath.path) else {\n",
+        "            continue\n",
+        "        }\n",
+        "        let data = try! Data(contentsOf: filePath, options: [])\n",
+        "        return [UInt8](data)\n",
+        "    }\n",
+        "    print(\"File not found: \\(path)\")\n",
+        "    exit(-1)\n",
+        "}\n",
+        "\n",
+        "/// Reads MNIST images and labels from specified file paths.\n",
+        "func readMNIST(imagesFile: String, labelsFile: String) -> (images: Tensor<Float>,\n",
+        "                                                           labels: Tensor<Int32>,\n",
+        "                                                           number: Int) {\n",
+        "    print(\"Reading data.\")\n",
+        "    let images = readFile(imagesFile).dropFirst(16).map(Float.init)\n",
+        "    let labels = readFile(labelsFile).dropFirst(8).map(Int32.init)\n",
+        "    let rowCount = labels.count\n",
+        "    let imageHeight = 28, imageWidth = 28\n",
+        "\n",
+        "    print(\"Constructing data tensors.\")\n",
+        "    return (\n",
+        "        images: Tensor(shape: [rowCount, 1, imageHeight, imageWidth], scalars: images)\n",
+        "            .transposed(withPermutations: [0, 2, 3, 1]) / 255, // NHWC\n",
+        "        labels: Tensor(labels),\n",
+        "        number: rowCount\n",
+        "    )\n",
+        "}"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "Urigd4SAAtQZ",
+        "colab_type": "text"
+      },
+      "source": [
+        "### Process the data\n",
+        "\n",
+        "Let's explore the format of the dataset before training the model. The following shows there are 60,000 images in the training set, and 10000 images in the test set:"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "mE76dHtHAumU",
+        "colab_type": "code",
+        "outputId": "a45ba6d3-bbb8-49f2-ceb8-87dffc181d84",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 119
+        }
+      },
+      "source": [
+        "let (trainImages, trainNumericLabels, numTrainExamples) = readMNIST(imagesFile: \"train-images-idx3-ubyte?raw=true\",\n",
```

Comment:
To avoid a line-wrap here, maybe you could move `imagesFile:` parameter to a new line just below `readMNIST(`. You could make it line up with the following `labelsFile:` parameter. Same with the readMNIST below this.